packages/compat/src/fixup-rules.js

```diff
@@ -0,0 +1,257 @@
+			},
+		});
+
+		// add passthrough methods
+		for (const [
+			contextMethodName,
+			sourceCodeMethodName,
+		] of removedMethodNames) {
+			newContext[contextMethodName] =
+				sourceCode[sourceCodeMethodName].bind(sourceCode);
+		}
+
+		// freeze just like the original context
+		Object.freeze(newContext);
+
+		/*
+		 * Create the visitor object using the original create() method.
+		 * This is necessary to ensure that the visitor object is created
+		 * with the correct context.
+		 */
+		const visitor = originalCreate(newContext);
+
+		/*
+		 * Wrap each method in the visitor object to update the currentNode
+		 * before calling the original method. This is necessary because the
+		 * methods like `getScope()` need to know the current node.
+		 */
+		for (const [methodName, method] of Object.entries(visitor)) {
+			/*
+			 * Node is the second argument to most code path methods,
+			 * and the third argument for onCodePathSegmentLoop.
+			 */
+			if (methodName.startsWith("on")) {
+				visitor[methodName] = (...args) => {
+					currentNode =
+						args[methodName === "onCodePathSegmentLoop" ? 2 : 1];
+
+					return method.call(visitor, ...args);
+				};
+
+				continue;
+			}
+
+			visitor[methodName] = (...args) => {
+				currentNode = args[0];
+
+				return method.call(visitor, ...args);
+			};
+		}
+
+		return visitor;
+	};
+
+	const newRuleDefinition = {
+		...(isLegacyRule ? undefined : ruleDefinition),
+		create: ruleCreate,
+	};
+
+	// cache the fixed up rule
+	fixedUpRuleReplacements.set(ruleDefinition, newRuleDefinition);
+	fixedUpRules.add(newRuleDefinition);
+
+	return newRuleDefinition;
+}
+
+/**
+ * Takes the given plugin and creates a new plugin with all of the rules wrapped
+ * to provide the missing methods on the `context` object.
+ * @param {FixupPluginDefinition} plugin The plugin to fix up.
+ * @returns {FixupPluginDefinition} The fixed-up plugin.
+ */
+export function fixupPluginRules(plugin) {
+	// first check if we've already fixed up this plugin
+	if (fixedUpPluginReplacements.has(plugin)) {
+		return fixedUpPluginReplacements.get(plugin);
+	}
+
+	/*
+	 * If the plugin has already been fixed up, or if the plugin
+	 * doesn't have any rules, we can just return it.
+	 */
+	if (fixedUpPlugins.has(plugin) || !plugin.rules) {
+		return plugin;
+	}
+
+	const newPlugin = {
+		...plugin,
+		rules: Object.fromEntries(
+			Object.entries(plugin.rules).map(([ruleId, ruleDefinition]) => [
+				ruleId,
+				fixupRule(ruleDefinition),
+			]),
+		),
+	};
+
+	// cache the fixed up plugin
+	fixedUpPluginReplacements.set(plugin, newPlugin);
+	fixedUpPlugins.add(newPlugin);
+
+	return newPlugin;
+}
+
+/**
+ * Takes the given configuration and creates a new configuration with all of the
+ * rules wrapped to provide the missing methods on the `context` object.
+ * @param {FixupConfigArray} configs The configuration to fix up.
+ * @returns {FixupConfigArray} The fixed-up configuration.
+ */
+export function fixupConfigRules(configs) {
```

Comment:
Should we maybe also support the case when `configs` is a config object, not an array?

Otherwise, maybe just document in README that it should be manually wrapped in an array  (`...fixupConfigRules([someConfig])`).

Reply:
Hmm yeah, probably a good idea to accept both to avoid confusion.